### DISCUSSION TITLE
how to read dwg text?

### DISCUSSION THREAD

**Discussion by farshadvalizade** [DomCR/ACadSharp#145] [Q&A]
Hi guys. I want to make a tool to read a dwg and extract it's content text into an array of data.
my dwg in autcad is :
![مخازن - حسینی_11-ژوئيه-2023_16-40-55](https://github.com/DomCR/ACadSharp/assets/63402203/a34c2f38-93c6-4187-84fd-84bfe4a4c059)
![مخازن - حسینی_11-ژوئيه-2023_16-41-15](https://github.com/DomCR/ACadSharp/assets/63402203/cdf5b3bd-8286-4878-a6d7-6945c5219fba)
the right table is my goal.what should I do?

**Comment by DomCR** [comment]
Hi @farshadvl,

Here is a quick guide of steps to read and manage information in a dwg file:

1. Use the `DwgReader` to read the file and obtain the object `CadDocument` - [example](https://github.com/DomCR/ACadSharp/blob/master/ACadSharp.Examples/ReaderExamples.cs)
2. All the entities in the model tab should be in `CadDocument.Entities` iterate through the list and look for `TextEntity` or `MText` to get the objects corresponding to TEXT and MTEXT in Autocad.
3. The text will be stored in the property `Value` for both types.

If there's anything you need, don't hesitate to ask.

Note: notice that the entity ACAD_TABLE is not implemented yet, if the text that you want to extract is in this type of entity, please open an issue and I'll fix it as soon as I can.

**Reply by farshadvalizade** [reply]
Tnx Sir. I have used this code for read:
`public static CadDocument ReadDwg(string file)
        {
            using (DwgReader reader = new DwgReader(file, onNotification))
            {
                CadDocument doc = reader.Read();
                return doc;
            }
        }`

and in my foreach loop I haven't seen any thing about TextEntity or MText.
where are they?
`OpenFileDialog ofd = new OpenFileDialog();
            ofd.ShowDialog();
            filepath = ofd.FileName;
            var doc = ReadDwg(filepath);
            foreach (var item in doc.Entities)
            {
               richTextBox1.Text = item ????
            }`

**Comment by DomCR** [comment]
TextEntity and MText are types, you can identify them using this:
```C#
	foreach (var item in doc.Entities)
	{
		string text = null;
		if(item is TextEntity texte)
		{
			text = texte.Value;
		}
		if(item is MText mtext)
		{
			text = mtext.Value;
		}
	}
```
But, before that, make sure that the objects that you are looking for are these types, with the images I cannot know if the text that you are trying to get is from a text object, an insert with attributes or a table.

**Reply by farshadvalizade** [reply]
Hi man I do that but it just returned number 3 in result.
can u see this file?
I think if we could convert dwg into dxf then we can simpler trace it.
Is there any way to convert this dwg into dxf ? 
[12-HEX-7020-1P1-PB(D)_001-REV01.zip](https://github.com/DomCR/ACadSharp/files/12045781/12-HEX-7020-1P1-PB.D._001-REV01.zip)

**Reply by DomCR** [reply]
If you want to convert to a dxf you can use this example:

```c#
	CadDocument doc;
	using (DwgReader reader = new DwgReader(_file))
	{
		doc = reader.Read();
	}

	using (DxfWriter writer = new DxfWriter(outputPath, doc, false))
	{
		writer.Write();
	}
```

**Comment by DomCR** [comment]
Hi @farshadvl,

I don't understand what you mean by "it just returned number 3 in result", but I ran a simple console app using your file and works fine for me:

![image](https://github.com/DomCR/ACadSharp/assets/46492487/c2fb4199-71dd-4f52-9a04-a8d6c364b8f1)

Here is the code:

```C#
static void Main(string[] args)
{
	CadDocument doc;
	using (DwgReader reader = new DwgReader(_file))
	{
		doc = reader.Read();
	}

	foreach (var item in doc.Entities)
	{
		if(item is TextEntity text)
		{
			Console.WriteLine($"TextEntity:{text.Value}");
		}
		else if(item is MText mtext)
		{
			Console.WriteLine($"MText:{mtext.Value}");
		}
	}
}
```

Inspecting the file I could see that some of the Text that I see in the image form AutoCAD are polylines:

![image](https://github.com/DomCR/ACadSharp/assets/46492487/3f4b6d56-013d-4b13-a1c3-662fd335242d)

This text won't be possible to extract as a string in the code.
